### PR TITLE
#4 BibTeX import/export

### PR DESCRIPTION
**Pull request by Dominic-DallOsto**
Work on enabling importing/exporting citations from BibTeX (#4).

Uses the Zotero Import Translator to convert from BibTeX to a Zotero item. We also get .ris support and more(?) for free - could change the option to be "import from file" or "import from clipboard"? Uses the `Zotero_File_Exporter` to export to a file, handling everything automatically.

Currently just reads from the clipboard, like Zotero's File > Import from Clipboard option. Would it be better to have a dialog box to paste the BibTeX into? To read from a file? Should we try to reconcile QIDs for all items on import? Otherwise it's quite tedious to do this item-by-item later.

I tested with the .ris citation list from [here](https://doi.org/10.1038/s41467-020-19757-y) - works OK apart from the problem that the raw data isn't complete.

- [x] Import from clipboard
- [x] Export support
- [x] Error handling
- [x] Translation support

Extensions:
- [x] Import menu (#124)
- [x] QID resolution (#38)
- [ ] Nicer export menu (#125)

### CODE REVIEW

**Comment by diegodlh** [comment]
Hi, @Dominic-DallOsto! Thank you so much for all the work you've been doing! Sorry I couldn't reply earlier. I have reviewed your changes and they look really good! I'll try to write a reply with comments tomorrow evening. Thanks so much again for your valuable contributions!!

**Comment by diegodlh** [comment]
> Work on enabling importing/exporting citations from BibTeX (#4).

This is so cool! This not only fixes #4, but also provides a workaround for #39 and #61, hence greatly simplifying adding citations and improving Cita's value. Thank you for all your work!

I'm sorry I couldn't reply to you earlier. I also regret not having had the time to implement automatic software tests yet (#30). I know it discourages making changes to the code and makes them so much more difficult :(


> We also get .ris support and more(?) for free - could change the option to be "import from file" or "import from clipboard"?

What about "Import citations from clipboard"?

> Would it be better to have a dialog box to paste the BibTeX into? To read from a file?

I think having a dialog box to paste into, with an option to read from a file and edit, would be ideal. But it may be more complex to code, and perfect is the enemy of good. I think "from clipboard" is OK for now, and we could open an issue to change it in the future. If you do implement the dialog box, though, what about having the option read "Import citations from file/clipboard", or just "Import citations"?

> Should we try to reconcile QIDs for all items on import? Otherwise it's quite tedious to do this item-by-item later.

That would be useful, but you may also consider these alternatives:
- Add a "Fetch QIDs for citations" option to the item menu on the Citations Pane. This was proposed in #38, but probably closed when the option to batch "Fetch QID(s)" for citing items was added to the item tree's Cita submenu.
- Auto fetch (citing and cited) QIDs before syncing citations with Wikidata (#36).

Either way, I would suggest doing only exact matches if batch fetching QIDs for two or more citations; i.e., not partial matches. This is the default behavior of the `Wikidata.reconcile` method for two or more items anyway.

> Uses the Zotero_File_Exporter to export to a file, handling everything automatically.

It's great to have other options than just BibTeX. However, I think that some translator options ("Export Notes", "Export Files", "Export Tags" and "Export Collections") that show for some formats may be confusing, as citations can't have notes, files or tags, or belong to specific collections (I'm not sure what the TEI translator option "Full TEI Document" means and whether it applies to citations, or what other options may be available for other export translators). What do you think?

I think the only way to remove these options would be to reimplement Zotero's `Zotero_File_Export.save` method, unfortunately. Maybe by changing translators' `displayOptions` array property [here](https://github.com/zotero/zotero/blob/9738d20558c13770ea0d58524c226a5ba60fe60d/chrome/content/zotero/fileInterface.js#L54), or by using a custom `exportOptions` dialog [here](https://github.com/zotero/zotero/blob/9738d20558c13770ea0d58524c226a5ba60fe60d/chrome/content/zotero/fileInterface.js#L55).

Alternatively, we may just skip format choice (maybe choose BibTeX by default), and copy to clipboard (instead of saving) for simplicity. But it would be so sad :'(

Depending on what choice we make, we should rename the option to either "Export citations to file" or "Export citations to clipboard as BibTeX".

> I tested with the .ris citation list from here - works OK apart from the problem that the raw data isn't complete.

Tested it myself. I can only say I'm so happy this is working so well!


At first I was hesitant about having methods that added citations without checking for duplicates. In fact, the first and simplest method to add citations, that is via the Citation Editor, does check for duplicates before adding a new citation. On the other hand, the "sync citations with Wikidata" method compares QIDs of local cited items and remote "cites work" statement objects; but it doesn't compare other fields (title, DOI, etc).

Hence, at some point, I realized that proactively making sure that no duplicate citations were created might be too difficult, and so I shifted to having a method (not yet implemented) to retroactively check for and merge duplicate citations (#37).

Although I wouldn't wait until having #37 closed to merge this pull request, I think it underscores the need to fix it.


I will add some in-line comments to your code now. Once again, thank you so much for this contribution! I can't highlight enough how important I think it is!!

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:495]: What about changing method's name to importCitationsFromClipboard?

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:496]: Please don't use `var`. Only `let` and `const`, as applicable.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:497]: Zotero's translation system may not be ready. Please add `await Zotero.Schema.schemaUpdatePromise;` above this line.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:502]: I see you have a "Translation support" task pending, so I guess you planned using `Wikicite.getString` here.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:519]: I think this one is redundant.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:530]: Maybe define the `citations` array at the beginning of the `try` block and at the end decide whether to show `Added citations` or `No citations were detected` depending on whether the `citations` array length is `>0`.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:514]: Maybe `jsonItems.forEach((jsonItem)` for clarity?

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:545]: Add `progress.close()`

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:527]: Maybe say maybe "X citations were imported" to emulate Zotero's import from clipboard behavior?

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:550]: `this.loadCitations()` should go before checking `this.citations` length.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:549]: Here and elsewhere, please use `this.citations` instead.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:496]: Please remove this message (and the corresponding export one) from the English locale.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:565]: I think it should work with `item.libraryID` either side of the assignment.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:573]: What about adding `exporter.name = Wikicite.getString(`wikicite.item-menu.export-file.filename');` so the default file name is something like "Exported citations" instead of "Exported items"?

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:551]: Please add `/* global Zotero_File_Exporter */` at the beginning of the file.

**Review by diegodlh — CHANGES_REQUESTED** [review]
This is my first time reviewing a non-trivial pull request and I wasn't sure whether to just push my suggestions. I didn't want to make a mess with your code. So I decided to make these in-line comments instead. But I'm willing to make any changes myself if you prefer.

File: src/sourceItemWrapper.js
Comment (diegodlh) [inline src/sourceItemWrapper.js:548]: Consider changing to `exportToFile`, `exportToClipboard` or `exportBibtexToClipboard`, depending on what we decide to do in the end.

File: src/sourceItemWrapper.js
Comment (Dominic-DallOsto) [reply src/sourceItemWrapper.js:519]: When removing these two lines and directly assigning jsonItem to the item field in new Citation, I receive an error that notes is not a valid field - in ItemWrapper.fromJSON. I tried deleting `notes` `seeAlso` and `attachments` from the jsonItem as in wikidata, but this still wasn't successful. I guess I'm just doing something wrong?

**Comment by Dominic-DallOsto** [comment]
> > Work on enabling importing/exporting citations from BibTeX (#4).
> 
> This is so cool! This not only fixes #4, but also provides a workaround for #39 and #61, hence greatly simplifying adding citations and improving Cita's value. Thank you for all your work!
> 
> I'm sorry I couldn't reply to you earlier. I also regret not having had the time to implement automatic software tests yet (#30). I know it discourages making changes to the code and makes them so much more difficult :(

No worries - thanks for developing Cita in the first place! 

> > Would it be better to have a dialog box to paste the BibTeX into? To read from a file?
> 
> I think having a dialog box to paste into, with an option to read from a file and edit, would be ideal. But it may be more complex to code, and perfect is the enemy of good. I think "from clipboard" is OK for now, and we could open an issue to change it in the future. If you do implement the dialog box, though, what about having the option read "Import citations from file/clipboard", or just "Import citations"?

Ok. For the moment I just changed it to import from clipboard. It should be easy to add another option for importing from a file. But yeah, a nicer dialog box can maybe be a later improvement...

> > Should we try to reconcile QIDs for all items on import? Otherwise it's quite tedious to do this item-by-item later.
> 
> That would be useful, but you may also consider these alternatives:
> 
>     * Add a "Fetch QIDs for citations" option to the item menu on the Citations Pane. This was proposed in [Batch fetch QID actions #38](https://github.com/diegodlh/zotero-cita/issues/38), but probably closed when the option to batch "Fetch QID(s)" for citing items was added to the item tree's Cita submenu.
> 
>     * Auto fetch (citing and cited) QIDs before syncing citations with Wikidata ([Offer to fetch QIDs before syncing citations #36](https://github.com/diegodlh/zotero-cita/issues/36)).
> 
> 
> Either way, I would suggest doing only exact matches if batch fetching QIDs for two or more citations; i.e., not partial matches. This is the default behavior of the `Wikidata.reconcile` method for two or more items anyway.

Ok, I'll have to think more about this - either of those solutions sounds good. Maybe it would be better not to reconcile automatically if a user wants use the extension offline.

> > Uses the Zotero_File_Exporter to export to a file, handling everything automatically.
> 
> It's great to have other options than just BibTeX. However, I think that some translator options ("Export Notes", "Export Files", "Export Tags" and "Export Collections") that show for some formats may be confusing, as citations can't have notes, files or tags, or belong to specific collections (I'm not sure what the TEI translator option "Full TEI Document" means and whether it applies to citations, or what other options may be available for other export translators). What do you think?
> 
> I think the only way to remove these options would be to reimplement Zotero's `Zotero_File_Export.save` method, unfortunately. Maybe by changing translators' `displayOptions` array property [here](https://github.com/zotero/zotero/blob/9738d20558c13770ea0d58524c226a5ba60fe60d/chrome/content/zotero/fileInterface.js#L54), or by using a custom `exportOptions` dialog [here](https://github.com/zotero/zotero/blob/9738d20558c13770ea0d58524c226a5ba60fe60d/chrome/content/zotero/fileInterface.js#L55).
> 
> Alternatively, we may just skip format choice (maybe choose BibTeX by default), and copy to clipboard (instead of saving) for simplicity. But it would be so sad :'(
> 
> Depending on what choice we make, we should rename the option to either "Export citations to file" or "Export citations to clipboard as BibTeX".

Yeah, I guess the extra options should be harmless, but definitely a bit confusing. I'll look at how the menu can be modified. An extra option to export BibTeX to clipboard might be convenient too, at the risk of having too many.


Thanks for the comments! I went through and hopefully addressed them all, except for creating a new Citation from a jsonItem. Feel free to make your own changes, but otherwise I'll work through the remaining changes.

File: src/sourceItemWrapper.js
Comment (diegodlh) [reply src/sourceItemWrapper.js]: I meant to anchor my comment to line 519, but now I see it's ambiguous what line I was referring to. I think the redundant line is `citation.target.item = newItem`. The citation's target item should be defined at instantiation on line 518. Does this make sense?

I'm entering a meeting now. I'll reply to your other comments in a moment. Thanks!!

**Comment by diegodlh** [comment]
> For the moment I just changed it to import from clipboard (...) a nicer dialog box can maybe be a later improvement...

Opened #124 to keep track of this.

> either of those solutions sounds good. Maybe it would be better not to reconcile automatically if a user wants use the extension offline.

I think I prefer having the button to batch fetch QIDs for citing items (#38). It would offer a workaround for both #36 and this (i.e., auto reconcile on import), and may be used in other cases as well. I've reopened #38.

> Yeah, I guess the extra options should be harmless, but definitely a bit confusing. I'll look at how the menu can be modified.

Again here I would support a good is better than perfect solution. Opened #125 to keep track of this.

> An extra option to export BibTeX to clipboard might be convenient too, at the risk of having too many.

I agree it may be too many options in the Citations Pane menu. I meant having export to clipboard as an alternative to export to file if we decided to reimplement `Zotero_File_Export.save` and wanted to avoid the complexity of using the file picker, etc.

If we redesign the export dialog to hide the confusing translator options mentioned above, we could have two buttons: one to save to a file and another to copy to the clipboard. Included this in #125.

> Thanks for the comments! I went through and hopefully addressed them all, except for creating a new Citation from a jsonItem. Feel free to make your own changes, but otherwise I'll work through the remaining changes.

Cool! I've replied to your comment about creating a new Citation from a jsonItem. Hope it helps. Thank you so much once again!!

**Comment by Dominic-DallOsto** [comment]
I just added the option to batch fetch QIDs for citations, and also for a single citation (not sure if needed?). The save handler for the citation `ItemWrapper`s didn't work, so I had to manually disable saving and them save the source item at the end. Not sure if this is correct?

I tested again with importing the .ris reference list from [here](https://doi.org/10.1038/s41467-020-19757-y) (Q108045069), the batch fetching the QIDs. This was a bit slow (5-10 s for 67 items) but worked. 

Would it be nice to have the progress message be "XX QIDs fetched" if more than one item is provided for reconciliation?

**Review by diegodlh — COMMENTED** [review]
I've just reviewed this. I sincerely apologize for the long delay. Everything looks quite good! I've made some minor changes which I'm going to push right away.

**Comment by diegodlh** [comment]
> I just added the option to batch fetch QIDs for citations, and also for a single citation (not sure if needed?). The save handler for the citation ItemWrappers didn't work, so I had to manually disable saving and them save the source item at the end. Not sure if this is correct?

Looks awesome! Thank you!! I've fixed the problem with the save handler for the citation target in acb1211aa6f0312ca083be4366e0e3b3b87b064e. Anyway, I've changed your code to use batch mode so the source item doesn't have to be saved each time a target QID is set.

> Would it be nice to have the progress message be "XX QIDs fetched" if more than one item is provided for reconciliation?

I like the idea. This would imply changing how `Wikidata.reconcile` handles the progress window. As you said, it would be for when more than one item is provided for reconciliation, and when approximate matches are disabled (default behavior for two or more items). I would suggest that you open a separate issue, though, so we can merge this without further changes. What do you think?


One question about importing citations. In `citation-importer/index.js` you wrote "todo: update with zotero/filePicker - discussion about this here: https://groups.google.com/g/zotero-dev/c/a1IPUJ2m_3s". Could you elaborate further on this point? Do you think we would have to address that before merging this PR? Or is this something we could fix later?

Finally, I'm going to push one or two more commits now, with Spanish and "documentation" (qqq) translations, and I think we should be ready to merge. What do you think? I would try and release a new version later this week, although I may wait a few days to roll out automatic updates, just in case.

I don't have an official way to communicate news about Cita. I could show an update message, but I would have to look into how to do it neatly, and I don't want to keep postponing this. I will just twit something. Do you use Twitter? Can I mention you in my twit?

**Review by diegodlh — APPROVED** [review]

**Comment by Dominic-DallOsto** [comment]
> I would suggest that you open a separate issue, though, so we can merge this without further changes. What do you think?

Yep, I added #128.

> One question about importing citations. In `citation-importer/index.js` you wrote "todo: update with zotero/filePicker - discussion about this here: https://groups.google.com/g/zotero-dev/c/a1IPUJ2m_3s". Could you elaborate further on this point? Do you think we would have to address that before merging this PR? Or is this something we could fix later?

In 2019 Zotero updated from using firefox 52 to firefox 60 as its base, which resulted in a number of changes as mentioned in that forum post. Specifically, the `nsIFilePicker.show()` method was replaced by `nsIFilePicker.open()`, but it was advised to use a newly added `FilePicker` zotero module instead. This is shown [here](https://github.com/zotero/zotero/commit/6f965251ed8eea160f95edaa8274c1bd786fafdc), with a backwards compatible version [here](https://github.com/jlegewie/zotfile/pull/424/commits/0feb00bb638fac792ceff5329dde5f8349b778c3#diff-c1242a57cfd1d6dbffa79f0bfeabb3891bbda60cb02d82751e76dc2bb5fe640e).

According to the comment from Emiliano Heyns in that google groups thread, it's not possible to use the new FilePicker when using webpack, because the module needs to be available at build time and this seemingly isn't the case for `'zotero/filePicker'`. I experienced the same - I couldn't get any code to compile that used the new `FilePicker`. Therefore, I used the same solution as Emiliano and stuck with the old `nsIFilePicker`. 

According to Dan Stillman in that same thread, at some point this will likely need to be changed to use the new `FilePicker`. But I imagine only when Zotero updates to a newer version of firefox than 60, and this doesn't seem like it will happen super soon. 

So, I think it's fine for now, but might need to be changed at some point in the future. We could make an issue with this information though?

> Finally, I'm going to push one or two more commits now, with Spanish and "documentation" (qqq) translations, and I think we should be ready to merge. What do you think? I would try and release a new version later this week, although I may wait a few days to roll out automatic updates, just in case.

Looks good! Adding citations by identifiers should be good to go too (might need a lot of the same small changes you made here though). But could also wait until after the dust has settled from this update - whatever you think.

> I don't have an official way to communicate news about Cita. I could show an update message, but I would have to look into how to do it neatly, and I don't want to keep postponing this. I will just twit something. Do you use Twitter? Can I mention you in my twit?

It seems possible to add release notes to the add-ons section in Zotero. I just updated my Better BibTeX extension and it had a section for release notes, but then they failed to load. So not sure if this works / would be easy to implement.
I don't use Twitter sorry, but feel free to mention me if that works. But no worries if not.

**Comment by diegodlh** [comment]
> the module needs to be available at build time and this seemingly isn't the case for 'zotero/filePicker'.

Thanks for the thorough explanation. I think I came across similar situations trying to import some Zotero's JSX components. I managed to import `zotero/filePicker` doing `import FilePicker from 'zotero@zotero/filePicker';` and adding  `'zotero@zotero/filePicker': 'commonjs zotero/modules/filePicker'` to the `externals` array in `webpack.config.js`. Could you give it a try, so we don't leave an open issue?

> Adding citations by identifiers should be good to go too (might need a lot of the same small changes you made here though). But could also wait until after the dust has settled from this update - whatever you think.

Yes! I'll try and check that now, if life allows. Thanks!

> It seems possible to add release notes to the add-ons section in Zotero. I just updated my Better BibTeX extension and it had a section for release notes, but then they failed to load. So not sure if this works / would be easy to implement.

I'm going to check that. Thanks!

**Comment by Dominic-DallOsto** [comment]
Thanks, that seems to work now. Do we want to also keep backwards compatibility with the FF52 version of Zotero like Zotfile does, or not worry about it given this version is ~2 years old?

**Review by diegodlh — APPROVED** [review]

**Comment by diegodlh** [comment]
Thank you very much for all your work on this, Dominic! And my apologies once again for taking so long to review and approve your changes.

I had to make one final tweak because of a change introduced in Zotero's beta channel.

> Do we want to also keep backwards compatibility with the FF52 version of Zotero like Zotfile does, or not worry about it given this version is ~2 years old?

I haven't checked compatibility with older versions of Zotero while developing Cita, and I'm sure there must be other things which are not compatible already. So let's not worry about this.

As I said before, I think your changes greatly improve Cita's usability and I will try and release a new version today or tomorrow including this and other changes. I think I'm gonna mark it as pre-release and wait to change the update.rdf until we have tested it a little bit.

Next big step should be addressing #30 so future contributions can be more straight-forward. Not an easy one, though :/

Thank you so much!!

**Pull request merged**
